src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -72,4 +92,203 @@ object RuleUtils {
+   * @return complementIndexPlan integrated plan of indexPlan and complementPlan
+   */
+  private def getComplementIndexPlan(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      originalPlan: LogicalPlan,
+      indexPlan: LogicalPlan): LogicalPlan = {
+    // 1) Replace the location of LogicalRelation with appended files
+    val complementIndexPlan = originalPlan transformUp {
+      case baseRelation @ LogicalRelation(
+            fsRelation @ HadoopFsRelation(location: PartitioningAwareFileIndex, _, _, _, _, _),
+            baseOutput,
+            _,
+            _) =>
+        // Set the same output schema with the index plan to merge them using BucketUnion
+        val updatedOutput =
+          baseOutput.filter(attr => index.schema.fieldNames.contains(attr.name))
+
+        val filesNotCovered = (location.allFiles
+          .map(f => FileInfo(f.getPath.toString, f.getLen, f.getModificationTime))
+          .toSet -- index.allSourceFiles).toSeq.map(f => new Path(f.name))
+
+        if (filesNotCovered.nonEmpty) {
+          val newLocation =
+            new InMemoryFileIndex(spark, filesNotCovered, Map(), None)
+          val newRelation =
+            fsRelation.copy(
+              location = newLocation,
+              dataSchema = StructType(index.schema.filter(baseRelation.schema.contains(_))))(
+              spark)
+          baseRelation.copy(relation = newRelation, output = updatedOutput)
+        } else {
+          baseRelation
+        }
+    }
+
+    if (!originalPlan.equals(complementIndexPlan)) {
+      // Remove sort order because we cannot guarantee the ordering of source files
+      val bucketSpec = index.bucketSpec.copy(sortColumnNames = Seq())
+      val attrs = complementIndexPlan.output.attrs.filter { attr =>
+        index.indexedColumns contains attr.name
+      }
+
+      // 2) Perform on-the-fly shuffle with the same partition structure of index
+      // so that we could avoid incurring shuffle whole index data at merge stage.
+      val shuffled = if (attrs.size != index.indexedColumns.size) {
+        // There should be a project node which excludes one or more indexedColumns.
+        // In this case, shuffle will be replaced to RoundRobinPartitioning which can cause
+        // wrong result issue. So shuffle should be located before the project node.
+        var projectSeen = false
+        val transformed = complementIndexPlan transformUp {
+          case project @ Project(_, child) =>
+            val childAttrs = child.output.attrs.filter { attr =>
+              index.indexedColumns contains attr.name
+            }
+            projectSeen = true
+            assert(childAttrs.size == index.indexedColumns.size)
+            project.copy(child = RepartitionByExpression(childAttrs, child, index.numBuckets))
+        }
+        assert(projectSeen)
+        transformed
+      } else {
+        RepartitionByExpression(attrs, complementIndexPlan, index.numBuckets)
```

Comment:
Question @imback82 , @sezruby do we want to `assert` the order of column names in attrs to match with bucket config?
I mean if bucketSpec says index columns [a, b] but in attrs we have columns [b, a]?

I may have missed if it was handled somewhere or not required at all.
Thanks.